src/components/Editor.tsx

```diff
@@ -1,15 +1,107 @@
+import {ARROW_LEFT, ARROW_RIGHT, BACKSPACE, META, ARROW_UP, ARROW_DOWN} from '../key';
+import {ISegment, SegmentType} from '../entities/segment';
+import {IEditorStateService, Subscriber} from '../services/editor-state.service';
+import {InsertTextAction} from '../state/insert-text.action';
+import {IEditorState} from '../state/editor.state';
+
+
+export default class Editor extends Component<any, IEditorState> {
+    private editorStateService = new IEditorStateService();
+
+    private readonly stateChangeSubscriber: Subscriber;
+
+    private editableRegion = createRef<HTMLDivElement>();
+
+    private isHoldingMetaKey = false;
+
+    constructor(props: any) {
+        super(props);
+        this.state = this.editorStateService.getState();
+
+        this.stateChangeSubscriber = () => this.setState(this.editorStateService.getState());
+        this.editorStateService.onStateChange(this.stateChangeSubscriber);
+    }
+
+    public componentWillUnmount() {
+        this.editorStateService.unSubscribe(this.stateChangeSubscriber);
+    }
 
-export class Editor extends Component<any, any> {
     render() {
         return (
-            <div
-                className={styles.Editor}
-                contentEditable
-                suppressContentEditableWarning
-            />
+            <div className={'Editor'}>
+                <div ref={this.editableRegion}
+                     className={'editable-region'}
+                     contentEditable
+                     suppressContentEditableWarning={true}
+                     onKeyDown={this.handleOnKeyDown}
+                     onKeyUp={this.handleOnKeyUp}
+                >
+                    {this.renderSegments()}
+                </div>
+            </div>
         );
     }
-}
+
+    private renderSegments() {
+        return this.state.segments.map(this.renderSegment);
+    }
+
+    private renderSegment = (segment: ISegment) => {
+        switch (segment.type) {
+            case SegmentType.Text:
+                return (
+                    <span data-segment-id={segment.idx}
+                          key={segment.idx}
+                    >
+                        {segment.content.join('')}
+                    </span>
+                );
+        }
+    };
+
+    private handleOnKeyDown = (event: KeyboardEvent<HTMLDivElement>) => {
+        switch (event.key) {
+            case BACKSPACE:
+                event.preventDefault();
+                return;
+            case ARROW_LEFT:
+                event.preventDefault();
+                return;
+            case ARROW_RIGHT:
+                event.preventDefault();
+                return;
+            case ARROW_UP:
+                event.preventDefault();
+                return;
+            case ARROW_DOWN:
+                event.preventDefault();
+                return;
+            case META:
+                event.preventDefault();
+                this.isHoldingMetaKey = true;
+                return;
```

Comment:
```suggestion
```

Reply:
Not in this PR.

Reply:
I think this is important to our implementation as it will give us a TODO list part, and it also helps us to prevent the default action of those keys.

Reply:
For this part. I believe that we need this part to prevent other types of key down events from affecting our project. This could also leave a TODO area for us.

Reply:
@yibolu Can we leave TODO for this. We should avoid adding all the features at once.